File: sui/ts/src/pyth.ts
import type {
  Transaction,
  TransactionArgument,
} from '@mysten/sui/transactions';
import type { SharedObjectInfo } from './utils';
import { addSharedObject, destroyZeroCoin, splitCoin } from './utils';
import { verifyVaa } from './wormhole';

/**
 * Updates price feeds using Pyth accumulator message
 * Returns an array of updated price info objects
 */
export function updatePriceFeeds(
  tx: Transaction,
  req: {
    pythPackage: string
    wormholePackage: string
    wormholeState: SharedObjectInfo
    pythState: SharedObjectInfo
    pythAccumulatorMessage: Uint8Array
    vaa: Uint8Array
    clock: TransactionArgument
    feeCoin: TransactionArgument
    feeCoinType: string
    priceInfoObjects: SharedObjectInfo[]
    updatePriceFee: bigint
  },
): TransactionArgument[] {
  // Verify VAA first
  const verifiedVaaArg = verifyVaa(tx, {
    wormholePackage: req.wormholePackage,
    wormholeState: req.wormholeState,
    clock: req.clock,
    vaa: req.vaa,
  });

  // Create pyth state arg
  const pythStateArg = addSharedObject(tx, req.pythState, false);

  // Get price updates hot potato
  let priceUpdatesHotPotatoArg = createAuthenticatedPriceInfosUsingAccumulator(
    tx,
    req,
    pythStateArg,
    verifiedVaaArg,
  );

  // Do price updates
  const priceInfoObjects: TransactionArgument[] = [];
  for (const priceInfoObject of req.priceInfoObjects) {
    const [priceInfoObjectArg, updatedPriceUpdatesHotPotatoArg]
      = updateSinglePricefeed(
        tx,
        req,
        pythStateArg,
        priceInfoObject,
        priceUpdatesHotPotatoArg,
      );

    priceInfoObjects.push(priceInfoObjectArg);
    priceUpdatesHotPotatoArg = updatedPriceUpdatesHotPotatoArg;
  }

  // Destroy zeroed fee coin
  destroyZeroCoin(tx, req.feeCoinType, req.feeCoin);

  // Destroy hot potato object
  destroyHotPotatoVector(tx, req, priceUpdatesHotPotatoArg);

  return priceInfoObjects;
}

/**
 * Creates authenticated price infos using accumulator
 * Returns HotPotatoVector<PriceInfo>
 */
function createAuthenticatedPriceInfosUsingAccumulator(
  tx: Transaction,
  req: {
    pythPackage: string
    pythAccumulatorMessage: Uint8Array
    clock: TransactionArgument
  },
  pythStateArg: TransactionArgument,
  verifiedVaaArg: TransactionArgument,
): TransactionArgument {
  const accumulatorMessageArg = tx.pure.vector('u8', req.pythAccumulatorMessage);

  return tx.moveCall({
    package: req.pythPackage,
    module: 'pyth',
    function: 'create_authenticated_price_infos_using_accumulator',
    typeArguments: [],
    arguments: [pythStateArg, accumulatorMessageArg, verifiedVaaArg, req.clock],
  });
}

/**
 * Updates a single price feed
 * Returns tuple of (price_info_object, updated_hot_potato)
 */
function updateSinglePricefeed(
  tx: Transaction,
  req: {
    pythPackage: string
    feeCoin: TransactionArgument
    feeCoinType: string
    updatePriceFee: bigint
    clock: TransactionArgument
  },
  pythStateArg: TransactionArgument,
  priceInfoObject: SharedObjectInfo,
  priceUpdatesHotPotatoArg: TransactionArgument,
): [TransactionArgument, TransactionArgument] {
  const feeForSingleFeed = splitCoin(
    tx,
    req.feeCoinType,
    req.feeCoin,
    req.updatePriceFee,
  );

  const priceInfoObjectArg = addSharedObject(tx, priceInfoObject, true);

  // Update price feed
  const extendedPriceUpdatesArg = tx.moveCall({
    package: req.pythPackage,
    module: 'pyth',
    function: 'update_single_price_feed',
    typeArguments: [],
    arguments: [
      pythStateArg,
      priceUpdatesHotPotatoArg,
      priceInfoObjectArg,
      feeForSingleFeed,
      req.clock,
    ],
  });

  return [priceInfoObjectArg, extendedPriceUpdatesArg];
}

/**
 * Destroys the hot potato vector
 */
function destroyHotPotatoVector(
  tx: Transaction,
  req: {
    pythPackage: string
  },
  hotPotatoVectorArg: TransactionArgument,
): void {
  const priceInfoType = `${req.pythPackage}::price_info::PriceInfo`;

  tx.moveCall({
    package: req.pythPackage,
    module: 'hot_potato_vector',
    function: 'destroy',
    typeArguments: [priceInfoType],
    arguments: [hotPotatoVectorArg],
  });
}

File: sui/rust/src/pyth.rs
use anyhow::Result;
use sui_sdk_types::{Address, Argument, StructTag, TypeTag};
use sui_transaction_builder::{Function, TransactionBuilder};

use crate::SharedObjectInfo;
use crate::utils::{add_bytes_to_tx, add_shared_object, destroy_zero_coin, split_coin};
use crate::wormhole::verify_vaas;

pub(crate) struct UpdatePriceFeeds<'a> {
    pub(crate) pyth_package: Address,
    pub(crate) wormhole_package: Address,
    pub(crate) wormhole_state: &'a SharedObjectInfo,
    pub(crate) pyth_state: &'a SharedObjectInfo,
    pub(crate) pyth_accumulator_message: &'a [u8],
    pub(crate) vaa: &'a [u8],
    pub(crate) clock: Argument,
    pub(crate) fee_coin: Argument,
    pub(crate) fee_coin_type: &'a TypeTag,
    pub(crate) price_info_objects: Vec<&'a SharedObjectInfo>,
    pub(crate) update_price_fee: u64,
}

/// Returns Vec<Argument<PriceInfoObject>>
pub(crate) fn update_price_feeds(
    tx: &mut TransactionBuilder,
    req: &UpdatePriceFeeds,
) -> Result<Vec<Argument>> {
    let verified_vaa_arg = verify_vaas(tx, req)?;

    // create pyth state arg
    let pyth_state_arg = add_shared_object(tx, req.pyth_state, false)?;

    // get price updates hot potato
    let mut price_updates_hot_potato_arg = create_authenticated_price_infos_using_accumulator(
        tx,
        req,
        pyth_state_arg,
        verified_vaa_arg,
    )?;

    // do price updates
    let mut price_info_objects = vec![];
    for price_info_object in &req.price_info_objects {
        let (price_info_object_arg, updated_price_updates_hot_potato_arg) =
            update_single_pricefeed(
                tx,
                req,
                pyth_state_arg,
                price_info_object,
                price_updates_hot_potato_arg,
            )?;

        price_info_objects.push(price_info_object_arg);
        price_updates_hot_potato_arg = updated_price_updates_hot_potato_arg;
    }

    // destroy zeroed fee coin
    destroy_zero_coin(tx, req.fee_coin_type, req.fee_coin)?;

    // destroy hot potato object
    destroy_hot_potato_vector(tx, req, price_updates_hot_potato_arg)?;

    Ok(price_info_objects)
}

/// Returns HowPotatoVector<PriceInfo>
fn create_authenticated_price_infos_using_accumulator(
    tx: &mut TransactionBuilder,
    req: &UpdatePriceFeeds,
    pyth_state_arg: Argument,
    verified_vaa_arg: Argument,
) -> Result<Argument> {
    let function = Function::new(
        req.pyth_package,
        "pyth".parse()?,
        "create_authenticated_price_infos_using_accumulator".parse()?,
        vec![],
    );

    let accumulator_message_arg = add_bytes_to_tx(tx, req.pyth_accumulator_message);

    Ok(tx.move_call(
        function,
        vec![
            pyth_state_arg,
            accumulator_message_arg,
            verified_vaa_arg,
            req.clock,
        ],
    ))
}

fn update_single_pricefeed(
    tx: &mut TransactionBuilder,
    req: &UpdatePriceFeeds,
    pyth_state_arg: Argument,
    price_info_object: &SharedObjectInfo,
    price_updates_hot_potato_arg: Argument,
) -> Result<(Argument, Argument)> {
    let fee_for_single_feed =
        split_coin(tx, req.fee_coin_type, req.fee_coin, req.update_price_fee)?;

    let price_info_object = add_shared_object(tx, price_info_object, true)?;
    // update price feed
    let function = Function::new(
        req.pyth_package,
        "pyth".parse()?,
        "update_single_price_feed".parse()?,
        vec![],
    );

    let extended_price_updates_arg = tx.move_call(
        function,
        vec![
            pyth_state_arg,
            price_updates_hot_potato_arg,
            price_info_object,
            fee_for_single_feed,
            req.clock,
        ],
    );

    Ok((price_info_object, extended_price_updates_arg))
}

fn destroy_hot_potato_vector(
    tx: &mut TransactionBuilder,
    req: &UpdatePriceFeeds,
    hot_potato_vector_arg: Argument,
) -> Result<()> {
    let function = Function::new(
        req.pyth_package,
        "hot_potato_vector".parse()?,
        "destroy".parse()?,
        vec![TypeTag::Struct(Box::new(StructTag {
            address: req.pyth_package,
            module: "price_info".parse()?,
            name: "PriceInfo".parse()?,
            type_params: vec![],
        }))],
    );

    tx.move_call(function, vec![hot_potato_vector_arg]);

    Ok(())
}
